// vite-plugin-monkey/dist/client/index.js
//#region src/client/window.ts
const monkeyWindow = document[__MONKEY_WINDOW_KEY__] ?? window;
//#endregion
//#region src/client/context.ts
const unsafeWindow = monkeyWindow.unsafeWindow;
const GM = monkeyWindow.GM;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_info
* @see https://violentmonkey.github.io/api/gm/#gm_info
*/
const GM_info = monkeyWindow.GM_info;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_log
* @available tampermonkey
*/
const GM_log = monkeyWindow.GM_log;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_getValue
* @see https://violentmonkey.github.io/api/gm/#gm_getvalue
*/
const GM_getValue = monkeyWindow.GM_getValue;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_getValues
* @see https://violentmonkey.github.io/api/gm/#gm_getvalues
*/
const GM_getValues = monkeyWindow.GM_getValues;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_setValue
* @see https://violentmonkey.github.io/api/gm/#gm_setvalue
*/
const GM_setValue = monkeyWindow.GM_setValue;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_setValues
* @see https://violentmonkey.github.io/api/gm/#gm_setvalues
*/
const GM_setValues = monkeyWindow.GM_setValues;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_deleteValue
* @see https://violentmonkey.github.io/api/gm/#gm_deletevalue
*/
const GM_deleteValue = monkeyWindow.GM_deleteValue;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_deleteValues
* @see https://violentmonkey.github.io/api/gm/#gm_deletevalues
*/
const GM_deleteValues = monkeyWindow.GM_deleteValues;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_listValues
* @see https://violentmonkey.github.io/api/gm/#gm_listvalues
*/
const GM_listValues = monkeyWindow.GM_listValues;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_addValueChangeListener
* @see https://violentmonkey.github.io/api/gm/#gm_addvaluechangelistener
*/
const GM_addValueChangeListener = monkeyWindow.GM_addValueChangeListener;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_removeValueChangeListener
* @see https://violentmonkey.github.io/api/gm/#gm_removevaluechangelistener
*/
const GM_removeValueChangeListener = monkeyWindow.GM_removeValueChangeListener;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_getResourceText
* @see https://violentmonkey.github.io/api/gm/#gm_getresourcetext
*/
const GM_getResourceText = monkeyWindow.GM_getResourceText;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_getResourceURL
* @see https://violentmonkey.github.io/api/gm/#gm_getresourceurl
*/
const GM_getResourceURL = monkeyWindow.GM_getResourceURL;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_addElement
* @see https://violentmonkey.github.io/api/gm/#gm_addelement
*/
const GM_addElement = monkeyWindow.GM_addElement;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_addStyle
* @see https://violentmonkey.github.io/api/gm/#gm_addstyle
*/
const GM_addStyle = monkeyWindow.GM_addStyle;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_openInTab
* @see https://violentmonkey.github.io/api/gm/#gm_openintab
*/
const GM_openInTab = monkeyWindow.GM_openInTab;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_getTab
* @available tampermonkey
*/
const GM_getTab = monkeyWindow.GM_getTab;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_saveTab
* @available tampermonkey
*/
const GM_saveTab = monkeyWindow.GM_saveTab;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_getTabs
* @available tampermonkey
*/
const GM_getTabs = monkeyWindow.GM_getTabs;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_registerMenuCommand
* @see https://violentmonkey.github.io/api/gm/#gm_registermenucommand
*/
const GM_registerMenuCommand = monkeyWindow.GM_registerMenuCommand;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_unregisterMenuCommand
* @see https://violentmonkey.github.io/api/gm/#gm_unregistermenucommand
*/
const GM_unregisterMenuCommand = monkeyWindow.GM_unregisterMenuCommand;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_notification
* @see https://violentmonkey.github.io/api/gm/#gm_notification
*/
const GM_notification = monkeyWindow.GM_notification;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_setClipboard
* @see https://violentmonkey.github.io/api/gm/#gm_setclipboard
*/
const GM_setClipboard = monkeyWindow.GM_setClipboard;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_xmlhttpRequest
* @see https://violentmonkey.github.io/api/gm/#gm_xmlhttprequest
*/
const GM_xmlhttpRequest = monkeyWindow.GM_xmlhttpRequest;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_download
* @see https://violentmonkey.github.io/api/gm/#gm_download
*/
const GM_download = monkeyWindow.GM_download;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_webRequest
* @available tampermonkey
*/
const GM_webRequest = monkeyWindow.GM_webRequest;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_cookie.list
* @see https://www.tampermonkey.net/documentation.php#api:GM_cookie.set
* @see https://www.tampermonkey.net/documentation.php#api:GM_cookie.delete
* @available tampermonkey
*/
const GM_cookie = monkeyWindow.GM_cookie;
/**
* @see https://www.tampermonkey.net/documentation.php#api:GM_audio.setMute
* @see https://www.tampermonkey.net/documentation.php#api:GM_audio.getState
* @see https://www.tampermonkey.net/documentation.php#api:GM_audio.addStateChangeListener
* @see https://www.tampermonkey.net/documentation.php#api:GM_audio.removeStateChangeListener
* @available tampermonkey
*/
const GM_audio = monkeyWindow.GM_audio;
//#endregion
export { GM, GM_addElement, GM_addStyle, GM_addValueChangeListener, GM_audio, GM_cookie, GM_deleteValue, GM_deleteValues, GM_download, GM_getResourceText, GM_getResourceURL, GM_getTab, GM_getTabs, GM_getValue, GM_getValues, GM_info, GM_listValues, GM_log, GM_notification, GM_openInTab, GM_registerMenuCommand, GM_removeValueChangeListener, GM_saveTab, GM_setClipboard, GM_setValue, GM_setValues, GM_unregisterMenuCommand, GM_webRequest, GM_xmlhttpRequest, monkeyWindow, unsafeWindow };
